overlord/devicestate/devicemgr.go

```diff
@@ -995,23 +995,62 @@ func (m *DeviceManager) Systems() ([]*System, error) {
 
 var ErrUnsupportedAction = errors.New("unsupported action")
 
+// Reboot triggers a reboot into the given systemLabel and mode.
+//
+// When called without a systemLabel and without a mode it will just
+// trigger a regular reboot.
+//
+// When called without a systemLabel but with a mode it will use
+// the current system to enter the given mode.
+//
+// Note that "recover" and "run" modes are only available for the
+// current system.
+func (m *DeviceManager) Reboot(systemLabel, mode string) error {
+	// most simple case: just reboot
+	if systemLabel == "" && mode == "" {
+		m.state.Lock()
+		defer m.state.Unlock()
+
+		logger.Noticef("restarting system")
+		m.state.RequestRestart(state.RestartSystemNow)
+		return nil
+	}
+
+	// no systemLabel means "current" so get the current system label
+	if systemLabel == "" {
+		systemMode := m.SystemMode()
+		currentSys, err := currentSystemForMode(m.state, systemMode)
+		if err != nil {
+			return fmt.Errorf("cannot get curent system: %v", err)
+		}
+		systemLabel = currentSys.System
+	}
+
+	return m.doSystemAction(systemLabel, SystemAction{Mode: mode})
+}
+
 // RequestSystemAction request provided system to be run in a given mode. A
 // system reboot will be requested when the request can be successfully carried
 // out.
 func (m *DeviceManager) RequestSystemAction(systemLabel string, action SystemAction) error {
 	if systemLabel == "" {
 		return fmt.Errorf("internal error: system label is unset")
 	}
+	return m.doSystemAction(systemLabel, action)
+}
 
+func (m *DeviceManager) doSystemAction(systemLabel string, action SystemAction) error {
 	if err := checkSystemRequestConflict(m.state, systemLabel); err != nil {
 		return err
 	}
 
 	systemMode := m.SystemMode()
+	// XXX: why do we ignore the error here?
```

Comment:
maybe that was correct at some point but I see at least  a path in currentSystemMode that can fail in various ways. Maybe the idea is not to stop people from doing any action even if we can't find the curren system, I think the error should be at least logged...

Reply:
This PR grew already a bit, happy to do this in a followup if you don't mind.

Reply:
anywway, this probably needs @bboozzoo reminding us why it is like this

Reply:
Maybe it is worth to at least log the error from `currentSystemForMode()`, but otherwise the idea was that inability to determine the current system should not block reinstall/recover requests if we're in trouble already.

With that being said, some lines further down there's this:
```
if systemMode == sysAction.Mode && systemLabel == currentSys.System {
```
which will panic when in recovery mode and the current system cannot be determined.

Reply:
I fixed the latter fwiw but didn't add a test yet

Reply:
Thanks for the fix.